add devices cgroup check as hard requirement

hqhq opened the pull request:
Devices cgroup is required for container security,  because without it a container will have unrestricted access to device nodes, libcontainer will take that as a hard requirement, we should warn users if they happen to have this umounted.

crosbymichael commented on pkg/sysinfo/sysinfo.go:
```diff
@@ -45,5 +45,17 @@ func New(quiet bool) *SysInfo {
+	// Check if Devices cgroup is mounted.
+	if _, err = cgroups.FindCgroupMountpoint("devices"); err != nil && !quiet {
+		if cgroups.IsNotFound(err) {
+			log.Warnf("Devices cgroup not found, it is required for container security.")
```
This should be a fatal not a warn.

crosbymichael commented on pkg/sysinfo/sysinfo.go:
```diff
@@ -5,7 +5,7 @@ import (
 	"path"
 
-	"github.com/Sirupsen/logrus"
+	log "github.com/Sirupsen/logrus"
```
Please change this back to `logrus`, we are doing it to help with imports

hqhq commented:
Updated. Thanks @crosbymichael

thaJeztah commented:
TestEventsFilterContainerID failed; possibly unrelated?

runcom commented:
@thaJeztah that test fails at random, try rebasing on master because I haven't seen it happening anymore

thaJeztah commented:
@runcom yes, I seemed to remember that being a flaky test. Thanks!

@hqhq can you try and rebase your PR? Thanks!

hqhq commented:
@thaJeztah this PR don't need rebase, it is still mergeable, if you want me rebase to trigger Jenkins again, it's done :)

thaJeztah commented:
@hqhq yes, I didn't think it needed a rebase, but it wouldnt hurt either :smile:. I see "all green" now, and that's what matters, thanks!

hqhq commented:
ping

cpuguy83 commented on pkg/sysinfo/sysinfo.go:
```diff
@@ -25,17 +25,15 @@ func New(quiet bool) *SysInfo {
-		_, err1 := ioutil.ReadFile(path.Join(cgroupMemoryMountpoint, "memory.limit_in_bytes"))
-		_, err2 := ioutil.ReadFile(path.Join(cgroupMemoryMountpoint, "memory.soft_limit_in_bytes"))
-		sysInfo.MemoryLimit = err1 == nil && err2 == nil
+		sysInfo.MemoryLimit = err == nil
```
Just for clarification: With this change, we assume that `limit_in_bytes` and `soft_limit_in_bytes` is there because `cgroup/memory` exists?

hqhq replied:
That's right, there are no extra kernel configs, one you mount memory cgroup, you'll have these files.

LK4D4 replied:
I don't unserstand this change then, because `err == nil` always here.

hqhq replied:
Oops, you are right, updated.

cpuguy83 commented on pkg/sysinfo/sysinfo.go:
```diff
@@ -45,5 +43,15 @@ func New(quiet bool) *SysInfo {
+
+	// Check if Devices cgroup is mounted, it is hard requirement for container security.
+	if _, err := cgroups.FindCgroupMountpoint("devices"); err != nil {
+		if cgroups.IsNotFound(err) {
```
perhaps since these are both Fatals, just do `log.Fatalf("error mounting devices cgroup: %v", err)` and remove the else.

hqhq replied:
Sure, updated, thanks.

hqhq committed:
add devices cgroup check and errors

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq committed:
simplify memory limit check

If memory cgroup is mounted, memory limit is always supported,
no need to check if these files are exist.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

thaJeztah commented:
@LK4D4 @cpuguy83 I see the PR was updated :)

LK4D4 commented on pkg/sysinfo/sysinfo.go:
```diff
@@ -22,20 +22,16 @@ func New(quiet bool) *SysInfo {
+		sysInfo.MemoryLimit = true
 
 		_, err = ioutil.ReadFile(path.Join(cgroupMemoryMountpoint, "memory.memsw.limit_in_bytes"))
 		sysInfo.SwapLimit = err == nil
```
here it is always true.

hqhq replied:
I don't get it, do you mean `sysInfo.SwapLimit` would always be true?
That's not right, `memory.memsw.limit_in_bytes` is not alway there even with memory cgroup mounted.

Or you mean something else?

LK4D4 replied:
`err == nil` always true here because you're in `else` branch for `err != nil`.

hqhq replied:
```
_, err = ioutil.ReadFile(path.Join(cgroupMemoryMountpoint, "memory.memsw.limit_in_bytes"))
sysInfo.SwapLimit = err == nil
```
that ioutil.ReadFile would refresh err, no?

jfrazelle replied:
I think it would

hqhq commented:
ping @LK4D4

jfrazelle commented:
also needs rebase sorry :(

hqhq commented:
@jfrazelle Rebased. 
And a slight change might make you happy @LK4D4

calavera commented:
code LGTM. Waiting for other maintainer to review.

cpuguy83 commented:
LGTM

thaJeztah commented:
Thanks @hqhq!
